File: cfms_fd/lib/ViewModel/Controller/AdminController/TeacherController.dart
import 'dart:convert';
import 'dart:io';
import 'package:file_picker/file_picker.dart';
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:http/http.dart' as http;

class TeacherController extends GetxController {
  // ⚠️ Modify base URL context as per your physical testing configuration
  final String baseUrl = "http://127.0.0.1:8000/api/admin";
  String authToken = "";

  // Core Reactive States
  var isLoading = false.obs;
  var isUploading = false.obs;
  var selectedFilePath = ''.obs;
  File? excelFile;

  // Datasets Storage Matrices
  var allTeachers = <dynamic>[].obs;
  var filteredTeachers = <dynamic>[].obs;

  // Dropdown Filtering Items
  var masterDepartments = <dynamic>[].obs;
  var departments = <String>[].obs;
  var selectedDepartment = ''.obs;

  Map<String, String> _getHeaders() {
    return {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      if (authToken.isNotEmpty) 'Authorization': 'Bearer $authToken',
    };
  }

  void initializeData(String token) {
    authToken = token;
    fetchDropdownFilters();
    fetchTeachers();
  }

  List<dynamic> _safeParseList(dynamic resBody) {
    try {
      var decoded = jsonDecode(resBody);
      if (decoded is Map && decoded.containsKey('data')) {
        return decoded['data'] as List<dynamic>;
      } else if (decoded is List) {
        return decoded;
      }
    } catch (e) {
      print("JSON extraction parsing fault: $e");
    }
    return [];
  }

  // ================= FETCH ALL DEPARTMENTS =================
  Future<void> fetchDropdownFilters() async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/department'), headers: _getHeaders()).timeout(const Duration(seconds: 15));
      if (response.statusCode == 200) {
        masterDepartments.assignAll(_safeParseList(response.body));
        List<String> depts = masterDepartments
            .map((e) => (e['dept_name'] ?? e['name'] ?? '').toString().trim())
            .where((s) => s.isNotEmpty)
            .toList();
        departments.assignAll(depts.toSet().toList());
      }
    } catch (e) {
      print("Error cascading dropdown layers: $e");
    }
  }

  // ================= FETCH ALL TEACHERS =================
  Future<void> fetchTeachers() async {
    isLoading.value = true;
    try {
      final response = await http.get(Uri.parse('$baseUrl/teachers'), headers: _getHeaders()).timeout(const Duration(seconds: 20));
      if (response.statusCode == 200) {
        allTeachers.assignAll(_safeParseList(response.body));
        applyFilters();
      } else {
        showSnackBar("Fetch Failure", "Could not query records from server configuration layout.", isError: true);
      }
    } catch (e) {
      showSnackBar("Connection Fault", "Network infrastructure pipeline breakdown: $e", isError: true);
    } finally {
      isLoading.value = false;
    }
  }

  // ================= DYNAMIC FILTER APPLICATION =================
  void onDepartmentChanged(String? val) {
    selectedDepartment.value = (val ?? '').trim();
    applyFilters();
  }

  void applyFilters() {
    if (allTeachers.isEmpty) {
      filteredTeachers.clear();
      return;
    }

    var dynamicFilter = allTeachers.where((teacher) {
      bool matchesDept = selectedDepartment.value.isEmpty;
      if (selectedDepartment.value.isNotEmpty && teacher['department'] != null) {
        String deptName = (teacher['department']['dept_name'] ?? teacher['department']['name'] ?? '').toString().trim();
        matchesDept = (deptName.toLowerCase() == selectedDepartment.value.toLowerCase());
      }
      return matchesDept;
    }).toList();

    filteredTeachers.assignAll(dynamicFilter);
  }

  // ================= BROWSE FILE PIPELINE =================
  Future<void> browseFile() async {
    try {
      FilePickerResult? result = await FilePicker.platform.pickFiles(
        type: FileType.custom,
        allowedExtensions: ['xlsx', 'xls', 'csv'],
      );

      if (result != null && result.files.single.path != null) {
        selectedFilePath.value = result.files.single.path!;
        excelFile = File(result.files.single.path!);
      } else {
        showSnackBar("Operation Halting", "Spreadsheet import matrix setup discarded by operator.", isError: true);
      }
    } catch (e) {
      showSnackBar("File Exception", "Critical operating system layer browsing fault: $e", isError: true);
    }
  }

  // ================= UPLOAD BULK EXCEL =================
  Future<void> importExcel() async {
    if (excelFile == null) {
      showSnackBar("Selection Empty", "Please choose a valid spreadsheet configuration template first.", isError: true);
      return;
    }

    isUploading.value = true;
    try {
      var request = http.MultipartRequest('POST', Uri.parse('$baseUrl/teachers/import'));
      request.headers.addAll({
        'Accept': 'application/json',
        if (authToken.isNotEmpty) 'Authorization': 'Bearer $authToken',
      });
      request.files.add(await http.MultipartFile.fromPath('file', excelFile!.path));

      var streamingStream = await request.send();
      var response = await http.Response.fromStream(streamingStream);

      dynamic parsedBody;
      try { parsedBody = jsonDecode(response.body); } catch (_) { parsedBody = null; }

      if (response.statusCode == 200) {
        showSnackBar("Dataset Synchronized", "Excel records populated inside database cluster smoothly.");
        selectedFilePath.value = '';
        excelFile = null;
        fetchTeachers();
      } else if (response.statusCode == 422) {
        String errorLog = parsedBody?['message'] ?? "Data alignment runtime validation failure.";
        if (parsedBody?['errors'] != null) {
          var validationTrack = parsedBody['errors'];
          if (validationTrack is Map) {
            validationTrack.forEach((k, v) => errorLog += "\n• ${(v is List) ? v.join(', ') : v}");
          }
        }
        _triggerCriticalModal("Validation Conflict (422)", errorLog);
      } else {
        _triggerCriticalModal("Server Error (${response.statusCode})", parsedBody?['message'] ?? "Internal processing fault.");
      }
    } catch (e) {
      _triggerCriticalModal("Transmission Exception", "Pipeline execution unexpected crush: $e");
    } finally {
      isUploading.value = false;
    }
  }

  // ================= EDIT / UPDATE RECORD =================
  Future<void> updateTeacher(int id, String name, String email, String password, String deptName) async {
    isLoading.value = true;
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/teachers/$id'),
        headers: _getHeaders(),
        body: jsonEncode({
          'teacher_name': name,
          'email': email,
          if (password.isNotEmpty) 'password': password,
          'department_name': deptName,
        }),
      );

      dynamic body = jsonDecode(response.body);
      if (response.statusCode == 200) {
        showSnackBar("Updated", "Teacher metadata attributes shifted inside schemas successfully.");
        fetchTeachers();
        Get.back();
      } else {
        showSnackBar("Mutation Refused", body['message'] ?? "Validation structure parsing mismatch.", isError: true);
      }
    } catch (e) {
      showSnackBar("Error", "Mutation payload transmission breakdown: $e", isError: true);
    } finally {
      isLoading.value = false;
    }
  }

  // ================= DELETE TEACHER =================
  Future<void> deleteTeacher(int id) async {
    try {
      final response = await http.delete(Uri.parse('$baseUrl/teachers/$id'), headers: _getHeaders());
      if (response.statusCode == 200) {
        showSnackBar("Dropped Record", "Identity structural links completely detached safely.");
        fetchTeachers();
      } else {
        showSnackBar("Erasure Terminated", "Could not execute removal pipelines context structural design.", isError: true);
      }
    } catch (e) {
      showSnackBar("Network Error", "Process execution failed: $e", isError: true);
    }
  }

  void showSnackBar(String title, String message, {bool isError = false}) {
    Get.snackbar(
      title, message,
      snackPosition: SnackPosition.BOTTOM,
      backgroundColor: isError ? Colors.redAccent.withOpacity(0.8) : Colors.green.withOpacity(0.8),
      colorText: Colors.white,
      margin: const EdgeInsets.all(16),
      barBlur: 10,
    );
  }

  void _triggerCriticalModal(String title, String msg) {
    showSnackBar(title, msg, isError: true);
    Get.defaultDialog(
      title: title,
      titleStyle: const TextStyle(color: Colors.redAccent, fontWeight: FontWeight.bold),
      content: SizedBox(
        height: 200, width: double.maxFinite,
        child: SingleChildScrollView(child: Text(msg, style: const TextStyle(fontSize: 12, fontFamily: 'monospace'))),
      ),
      textConfirm: "Acknowledge",
      confirmTextColor: Colors.white,
      buttonColor: Colors.redAccent,
      onConfirm: () => Get.back(),
    );
  }
}

File: cfms_bd/app/Http/Controllers/Admin/TeacherController.php
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Controllers\Controller;
use App\Models\Teacher;
use App\Models\User;
use App\Models\Department;
use App\Imports\TeacherImport;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Hash;
use Illuminate\Support\Facades\Validator;
use Maatwebsite\Excel\Facades\Excel;

class TeacherController extends Controller
{
    /**
     * Display a listing of teachers with cascaded relations.
     */
    public function index(Request $request)
    {
        $query = Teacher::with(['user', 'department']);

        // Screen department filter pipeline matching strictly 'dept_name'
        if ($request->has('department') && !empty($request->department)) {
            $deptFilter = $request->department;
            $query->whereHas('department', function($q) use ($deptFilter) {
                $q->where('dept_name', $deptFilter);
            });
        }

        $teachers = $query->latest()->get();

        return response()->json([
            'success' => true,
            'data' => $teachers
        ], 200);
    }

    /**
     * Update the specific teacher records within pipeline transactions.
     */
    public function update(Request $request, $id)
    {
        $teacher = Teacher::find($id);
        if (!$teacher) {
            return response()->json(['message' => 'Teacher record not found.'], 404);
        }

        $validator = Validator::make($request->all(), [
            'teacher_name' => 'required|string|max:255',
            'email' => 'required|email|unique:users,email,' . $teacher->user_id,
            'password' => 'nullable|min:6',
            'department_name' => 'required|string',
        ]);

        if ($validator->fails()) {
            return response()->json([
                'message' => 'Validation failure on backend pipelines.',
                'errors' => $validator->errors()
            ], 422);
        }

        $department = Department::where('dept_name', $request->department_name)->first();

        if (!$department) {
            return response()->json([
                'message' => 'Validation failure on backend pipelines.',
                'errors' => ['department_name' => ['Targeted department does not exist in databases.']]
            ], 422);
        }

        DB::beginTransaction();
        try {
            // Update profile
            $teacher->update([
                'teacher_name' => $request->teacher_name,
                'dept_id' => $department->id,
            ]);

            // Sync account properties
            $userData = ['email' => $request->email];
            if ($request->filled('password')) {
                $userData['password'] = Hash::make($request->password);
            }
            
            User::where('id', $teacher->user_id)->update($userData);

            DB::commit();
            return response()->json(['message' => 'Teacher details updated successfully.'], 200);

        } catch (\Exception $e) {
            DB::rollBack();
            return response()->json(['message' => 'Failed to save data record.'], 500);
        }
    }

    /**
     * Remove the teacher profile and cascade delete user credentials safely.
     */
    public function destroy($id)
    {
        $teacher = Teacher::find($id);
        if (!$teacher) {
            return response()->json(['message' => 'Teacher record not found.'], 404);
        }

        try {
            // Cascade delete user profile. Migration triggers teacher removal automatically
            User::destroy($teacher->user_id);

            return response()->json(['message' => 'Teacher record dropped successfully.'], 200);
        } catch (\Exception $e) {
            return response()->json(['message' => 'Could not execute deletion pipeline.'], 500);
        }
    }

    /**
     * Upload and parse spreadsheet datasets bulk matrix.
     */
    public function import(Request $request)
    {
        $validator = Validator::make($request->all(), [
            'file' => 'required|mimes:xlsx,xls,csv|max:10240',
        ]);

        if ($validator->fails()) {
            return response()->json([
                'message' => 'Data validation errors occurred during import.',
                'errors' => $validator->errors()
            ], 422);
        }

        DB::beginTransaction();
        try {
            Excel::import(new TeacherImport, $request->file('file'));
            DB::commit();

            return response()->json(['message' => 'Excel spreadsheet datasets imported successfully.'], 200);
        } catch (\Illuminate\Validation\ValidationException $e) {
            DB::rollBack();
            return response()->json([
                'message' => 'Data validation errors occurred during import.',
                'errors' => $e->errors()
            ], 422);
        } catch (\Exception $e) {
            DB::rollBack();
            
            // Shows dynamic descriptive validation or SQL mismatches straight to Flutter
            return response()->json([
                'message' => 'Import Error: ' . $e->getMessage(),
                'errors' => ['file' => [$e->getMessage()]]
            ], 500);
        }
    }
}